### PR TITLE
fix: make cancel field required or optional

### PR DESCRIPTION
**Pull request by Chiranjeev-droid**
## What does this PR do?

This pr add an option to make the cancelation reason required / optional within event types.

- Fixes #15977 

 Loom Video: https://www.loom.com/share/9666b8c46d6a403a86baa5eb3a78118b?sid=92314e30-877f-4ee0-bef3-731ecd942b48

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] Na
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
Go to advanced tab of event types and make the cancellation reason field optional / required and change the default label or placeholder if needed for the  cancel reason field.
Then create a booking and try cancelling it , you will then be able to see the changes.

### CODE REVIEW

**Comment by Chiranjeev-droid** [comment]
@anikdhabal i have made the suggested changes.

**Comment by Chiranjeev-droid** [comment]
i have made the required changes @anikdhabal .

File: packages/features/bookings/lib/getBookingFields.ts
Comment (anikdhabal) [inline packages/features/bookings/lib/getBookingFields.ts:291]: @Chiranjeev-droid thanks for the work. It works, but the cancellation reason is not part of the booking fields. I’m not sure we can consider this a viable fix

Comment (Chiranjeev-droid) [reply packages/features/bookings/lib/getBookingFields.ts:291]: if needed , we can create a field in database saying cancellationReasonNeeded inside the eventtype and just add a toogle on the advanced tab. Based on the boolean value , we can make the reason required or optional but then user wont be able to customize the cancellationreason field.

File: packages/features/bookings/lib/handleCancelBooking.ts
Comment (Amit91848) [inline packages/features/bookings/lib/handleCancelBooking.ts:184]: Please avoid type gymnastics here 😅. We have `eventTypeBookingFields` schema is `zod-utils`. You should use that instead so that you know properly what you are dealing with.

File: apps/web/components/booking/CancelBooking.tsx
Comment (Amit91848) [inline apps/web/components/booking/CancelBooking.tsx:54]: Same here. I see know you have already imported eventTypeBookingFields. You should use 
```tsx
const parsedFields = eventType?.bookingFields ? eventTypeBookingFields.parse(eventType.bookingFields) : null;
parsedFields.find(() => ) .....or something similar that suits your needs.
```

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Left some suggestions code-wise @Chiranjeev-droid . Will get back to you about how we want to proceed with making Cancellation Reason required

File: packages/features/bookings/lib/getBookingFields.ts
Comment (anikdhabal) [reply packages/features/bookings/lib/getBookingFields.ts:291]: Yeah, we got approval from the team. Let’s make it a setting toggle in the advanced tab instead of a booking question

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Requested changes for visibility ->  change it to a settings toggle instead. https://github.com/calcom/cal.com/pull/16546#discussion_r1750740101

**Review by Amit91848 — APPROVED** [review]
LGTM now! Thanks for you contribution @Chiranjeev-droid ! 

Will wait for another review before merging

File: apps/web/components/booking/CancelBooking.tsx
Comment (anikdhabal) [inline apps/web/components/booking/CancelBooking.tsx:57]: This is not needed.` bookingCancellationReasonRequired` is not an optional property; it's either true or false. We can use it directly. Currently, we use unknown as the type for the event and booking props in this component. If we fix that, we won't need to do any extra work. eg:- `eventType: PageProps["eventType"]`;

File: apps/web/components/booking/CancelBooking.tsx
Comment (anikdhabal) [inline apps/web/components/booking/CancelBooking.tsx:106]: Does this work? `cancellationReason` is only a string type, so how does it work with `!cancellationReason`? Instead, we should check the length.

**Review by anikdhabal — CHANGES_REQUESTED** [review]
Left few comments

File: apps/web/components/booking/CancelBooking.tsx
Comment (Amit91848) [reply apps/web/components/booking/CancelBooking.tsx:106]: Empty string ("") is considered as undefined

Screenshot 2024-09-23 105351

File: apps/web/components/booking/CancelBooking.tsx
Comment (Amit91848) [reply apps/web/components/booking/CancelBooking.tsx:57]: I think eventtype is set as unknown for some reason. Not completely sure about it which is why I did not change it.

cc @Udit-takkar

**Pull request closed without merging**
